File: choeunwoo/week3/boj_18258.c
#include <stdio.h>
#include <string.h>

#define MAX 2000000


int queue[MAX];
int front_idx = 0;
int back_idx = 0;

void push(int x);
int pop();
int size();
int empty();
int front();
int back();

int main() {
    int N;
    char cmd[10];
    int x;

    scanf("%d", &N);

    while (N--) {
        scanf("%s", cmd);

        if (strcmp(cmd, "push") == 0) {
            scanf("%d", &x);
            push(x);
        }
        else if (strcmp(cmd, "pop") == 0) {
            printf("%d\n", pop());
        }
        else if (strcmp(cmd, "size") == 0) {
            printf("%d\n", size());
        }
        else if (strcmp(cmd, "empty") == 0) {
            printf("%d\n", empty());
        }
        else if (strcmp(cmd, "front") == 0) {
            printf("%d\n", front());
        }
        else if (strcmp(cmd, "back") == 0) {
            printf("%d\n", back());
        }
    }

    return 0;
}

void push(int x) {
    queue[back_idx++] = x;
}
int pop() {
    if (front_idx == back_idx) return -1;
    return queue[front_idx++];
}
int size() {
    return back_idx - front_idx;
}
int empty() {
    return front_idx == back_idx ? 1 : 0;
}
int front() {
    if (front_idx == back_idx) return -1;
    return queue[front_idx];
}
int back() {
    if (front_idx == back_idx) return -1;
    return queue[back_idx - 1];
}

File: choitaejin/week3/boj_18258.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

class Node {
  constructor(item) {
    this.item = item;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  getSize() {
    return this.size;
  }

  enqueue(item) {
    const node = new Node(item);

    if (!this.head) {
      this.head = node;
      this.tail = node;
    }
    else {
      this.tail.next = node;
      this.tail = node;
    }

    this.size++;
  }

  isEmpty() {
    return this.size === 0 ? 1 : 0;
  }

  dequeue() {
    if (this.isEmpty()) {
      return -1;
    }

    const item = this.head.item;
    this.head = this.head.next;
    this.size--;

    if (this.isEmpty()) {
      this.tail = null;
    }

    return item;
  }


  front() {
    if (this.isEmpty()) {
      return -1;
    }
    else {
      return this.head.item;
    }
  }

  back() {
    if (this.isEmpty()) {
      return -1;
    }
    else {
      return this.tail.item;
    }
  }

}

let input = [];
let lineCnt = 0;
let n = 0;
const queue = new Queue();
let res = [];

rl.on('line', (line) => {
  if (lineCnt === 0) {
    n = parseInt(line);
    lineCnt++;
  } else {
    input.push(line);
    lineCnt++;
  }

  if (lineCnt > n) {
    rl.close();
  }
}).on('close', () => {
  for (let i = 0; i < n; i++) {
    const parts = input[i].split(' ');
    const cmd = parts[0];

    switch (cmd) {
      case 'push':
        queue.enqueue(parts[1]);
        break;
      case 'pop':
        res.push(queue.dequeue());
        break;
      case 'size':
        res.push(queue.getSize());
        break;
      case 'empty':
        res.push(queue.isEmpty());
        break;
      case 'front':
        res.push(queue.front());
        break;
      case 'back':
        res.push(queue.back());
        break;
      default:
        break;
    }
  }

  console.log(res.join('\n'));
  process.exit();
});
